063-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,208 @@
+
+1. The new reporters are added as an additional options to the Strimzi custom resource, but we keep supporting the JMX Exporter and use it in our examples as the main thing (with possibly some new examples of CRs and dashboards prepared as well, but not as the main thing)
+
+2. We switch to the new reporters as the main example. We keep supporting JMX Exporter but deprecate it.
+
+3. Later, once ZooKeeper is not supported anymore, we might decide to drop the JMX Exporter support completely.
+
+### strimzi-kafka-oauth
+
+The OAuth plugin exposes the `strimzi.oauth.metric.reporters` configuration to let users specify a metrics reporter. Today the OAuth plugin automatically uses `org.apache.kafka.common.metrics.JmxReporter` if that configuration is not set. To switch to the new reporter, users should set `strimzi.oauth.metric.reporters` to `KafkaPrometheusMetricsReporter`.
+
+Note that directly instantiating metric reporters in the plugin is a workaround as Kafka currently does not provide a way for plugins to register metrics. [KIP-877](https://cwiki.apache.org/confluence/display/KAFKA/KIP-877%3A+Mechanism+for+plugins+and+connectors+to+register+metrics), currently in discussion, aims at solving this issue. Once this KIP is accepted, we should update the plugin to use this mechanism.
+
+### strimzi-kafka-bridge
+
+To function, the bridge instantiates Kafka clients. Today it has its own custom mechanism to retrieve metrics from the Kafka clients via JMX and expose them to Prometheus.
+
+We can update the bridge to use metric reporters to retrieve metrics from Kafka clients and still keep it's own mechanism to expose them to Prometheus (as it's also exposing its own metrics). To switch behavior, I propose introducing a new configuration to the bridge `metricsMode`/`KAFKA_BRIDGE_METRICS_MODE` which would initially default to `jmx` to keep the current behavior. If set to `reporter`, the bridge would set the `metric.reporters` configuration on all Kafka clients it starts to `KafkaPrometheusMetricsReporter` and retrieve their metrics via the Prometheus metrics registry. It should also set `prometheus.metrics.reporter.listener` to `disabled` so `KafkaPrometheusMetricsReporter` instances don't start their own HTTP endpoint.
+
+### kafka-quotas-plugin
+
+This plugin creates metrics by using the Yammer metrics library. `com.yammer.metrics.Metrics` registers metrics in the default Yammer registry and automatically exports metrics via JMX.
+
+Like strimzi-kafka-oauth, this plugin should be updated once [KIP-877](https://cwiki.apache.org/confluence/display/KAFKA/KIP-877%3A+Mechanism+for+plugins+and+connectors+to+register+metrics) is available in Kafka. 
+
+In the meantime, `YammerPrometheusMetricsReporter` will retrieve metrics from both the Kafka Yammer registry and the default Yammer registry to get all Yammer metrics.
+
+## Compatibility
+
+Differences with jmx_exporter metrics:
```

Comment:
I'm not clear what I'm seeing in this section. Are we listing the limitations with the proposed approach ("the reporter") against using the jmx_exporter to collect metrics?

Reply:
This section lists the differences in the metrics that end up in Prometheus using jmx_exporter and the reporters. 

While the outputs are pretty close, there are some differences and I wanted to highlight them all. It's not really that the output from one is superior to the other, they are just different.